src/cleanup.ml

```diff
@@ -2,17 +2,50 @@ open Lwt.Infix
 
 module Xen_api = Xen_api_lwt_unix
 
-module StringSet = Set.Make(String)
+module Local_xapi_session = struct
+  let wait_for_xapi_and_login () =
+    let rpc = Xen_api.make Consts.xapi_unix_domain_socket_uri in
+    let rec loop () =
+      Lwt.catch
+        (fun () -> Xen_api.Session.login_with_password ~rpc ~uname:"" ~pwd:"" ~version:"1.0" ~originator:"xapi-nbd")
+        (fun e ->
+           Lwt_log.warning_f "Failed to log in via xapi's Unix domain socket: %s; retrying in %f seconds" (Printexc.to_string e) Consts.wait_for_xapi_retry_delay_seconds >>= fun () ->
+           Lwt_unix.sleep Consts.wait_for_xapi_retry_delay_seconds >>= fun () ->
+           loop ()
```

Comment:
What about limiting the number of retries as well?

Reply:
Hmm, so you reckon that's also necessary in addition to the timeout? :thinking:

Reply:
I don't know. I was asking for your opinion. Is it worth trying forever or we should back out and fail after few minutes of attempts?

Reply:
Actually, the number of retries is limited because we have a timeout below, which is 120 seconds, and we retry every 4 seconds, so the max no of retries is 30.

Reply:
Yes, it was more about doing one or the other. As we discussed let's leave it as it is but increase the overall timeout to `300.0`

Reply:
Done - now the timeout is `300` seconds.

Reply:
In general it's a bad idea to retry forever.   Having software just "hang" is very frustrating for the user.  Once it is clear that something isn't going to work you should stop retrying and escalate the problem to the next level of process management logic until eventually the user gets a message.